Add blog post for Ensuring Secret Pulled Images (#49999)
kubernetes/website: content/en/blog/_posts/2025-04-09-ensure-secret-pulled-images/index.md

---
layout: blog

> title: "Image Pull Policy the way you always thought it worked!"

Author: Updating.

date: ???
draft: true
slug: ensure-secret-pulled-images-alpha
author: >
  [Ben Petersen](https://github.com/benjaminapetersen) (Microsoft),
  [Stanislav Láznička](https://github.com/stlaz) (Microsoft)
---

## Image Pull Policy the way you always thought it worked!

Some things in Kubernetes are surprising, and the way `imagePullPolicy` behaves might

> be one of them. Given Kubernetes is all about running pods, it may be surprising

Reviewer: vs using surprising in two sentences in a row...

```suggestion
be one of them. Given Kubernetes is all about running pods, it may be eye-opening
```

Author: Agree, updating!

to learn that there has been a caveat to restricting pod access to authenticated images for
over 10 years in the form of [issue 18787](https://github.com/kubernetes/kubernetes/issues/18787)!
It is an exciting release when you can resolve a ten year old issue.

{{< note >}}
Throughout this blog post, the term "pod credentials" will be used often. In this context,
the term generally encapsulates the authentication material that is available to a pod
to authenticate a container image pull.
{{< /note >}}

## IfNotPresent, even if I'm not supposed to have it

The gist of the problem is that the `imagePullPolicy: IfNotPresent` strategy has done

> exactly its job. *Pod A* in a *Namespace X* is scheduled to a *Node 1*, and requires

Reviewer: ```suggestion
its job, however when an image requires credentials ensuring a pod has access can be circumvented with the `IfNotPresent` policy. For example: *Pod A* in a *Namespace X* is scheduled to a *Node 1*, and requires
```

Author: Reworked this, will push.

The Kubelet will utilize the credentials from *Secret 1* as supplied by *Pod A*
and it pulls *container image Foo* from the registry. This is exactly what you expect.

But, if *Pod B* in *Namespace Y* happens to also be scheduled to *Node 1*, funny things
can happen. *Pod B* may reference the same private image, specifying the `IfNotPresent`
image pull policy. When the Kubelet tries to run the pod, it honors the `IfNotPresent`
policy, it sees that the *image Foo* is already present locally, and will
provide *image Foo* to *Pod B*. Surely that is ok since the container image is already
present, right?

{{< figure
    src="ensure_secret_image_pulls.svg"
    caption="Using a private image pulled by a different pod"
    alt="Illustration of the process of two pods trying to access a private image, the first one with a pull secret, the second one without it"
>}}

Wrong. While `IfNotPresent` should not pull *image Foo* if it is already present
on the node, it is an incorrect security posture to allow all pods scheduled
to a node to have access to that private image, discarding the fact that these
pods would otherwise never be able to pull that image from the remote repository

> in the first place.

Reviewer: ```suggestion
in the first place. Thus, especially in multi-tenant environments, the `imagePullPolicy: Always` has been the proper security posture.
```

## IfNotPresent, but only if I am supposed to have it

In Kubernetes 1.33, we - SIG Auth and SIG Node - are finally addressing this (really old) problem and
getting the verification right! As expected, if an image is not present, the Kubelet will
attempt to pull the image. The credentials each pod supplies will be utilized for this
task. This matches behavior prior to 1.33.

If the image is present, this is where the behavior changes.
The Kubelet will now verify the pod's credentials before allowing the pod to use
the image.

> Performance and service stability have been a consideration while designing the feature.

Reviewer: ```suggestion
Performance and service stability have been a consideration while designing this feature.
```

Author: 👍

Pods utilizing the same credential will not be required to re-authenticate. This is
also true in case the credentials sourced from the same Kubernetes Secret object rotated.

## Never pull, but prove you can use it!

The `imagePullPolicy: Never` option does not fetch images. Ever. However, if the
container image is already present on the node, the pod's credentials to use the
image may still require verification.

Just like before, pods utilizing the same credential will not be required to re-authenticate.
If the pod is not able to supply credentials previously used to successfully pull an
image, though, the `Never` image pull policy will make it fail because without
the ability to pull images, the pod just has no way to access the image.

## Always?

The `imagePullPolicy: Always` has always worked as intended. Each time an image
is requested, the request goes to the registry and there will be an authentication
check.

In the past, forcing the `Always` image pull policy via pod admission was the only way to ensure
that your private container images didn't get reused by other pods on nodes which already pulled the images.

> Fortunately, only the image manifest usually gets pulled in these cases instead

Reviewer: ```suggestion
Fortunately, only the image manifests usually gets pulled in these cases instead
```

Author: Reworked.

> of the whole image. However, this means that whenever this check is done (during a new rollout,

Reviewer: ```suggestion
of the whole image (the large layers making up the image contents are cached by SHA). However, this means that whenever this check is done (during a new rollout,
```

Author: Reworked.

     the already-pulled image.
  1. If the credentials or their source Secret are not found in the records of
     previous successful pulls for that image, the Kubelet will attempt to use
     these new credentials to request a pull from the remote registry.

## Try it out

In Kubernetes 1.33 we shipped the alpha version of this feature. To give it a spin,
enable the `KubeletEnsureSecretPulledImages` feature gate for your 1.33 Kubelets.

You can learn more about the feature and additional optional configuration on the
[concept page for Images](/docs/concepts/containers/images/#ensureimagepullcredentialverification)
in the official Kubernetes documentation.

## What's next?

In future releases, there are a couple of areas for us to focus on:
1. Make the feature work together with [Projected service account tokens for Kubelet image credential providers](https://kep.k8s.io/4412) which
   adds a new, workload-specific source of image pull credentials.
1. Write a benchmarking suite to be able to measure the performance of the feature,
   and to be able to measure the performance impact of any future changes to it.

> 1. Implement an in-memory caching layer so that we don't need to read files for each image pull request.

Reviewer: ```suggestion
1. Implement an in-memory caching layer so that we don't need to read files for each image pull request.
1. Add support for credential expirations, thus forcing previously validated credentials to be re-authenticated.  
```

Author: 👍

---
layout: blog
title: "Image Pull Policy the way you always thought it worked!"
date: 2025-04-08

> draft: false

Reviewer 2: ```suggestion
draft: true
```

See https://kubernetes.io/docs/contribute/blog/article-submission/#article-scheduling for rationale.

Author: Set `draft` and also updated `date: ???` as it looks like the date is updated later on a `publish` PR.

slug: ensure-secret-pulled-images-alpha
author: >
  [Ben Petersen](https://github.com/benjaminapetersen) (Microsoft),
  [Stanislav Láznička](https://github.com/stlaz) (Microsoft)
---

## Image Pull Policy the way you always thought it worked!

Some things in Kubernetes are surprising, and the way ImagePullPolicy behaves might
be one of them. Given Kubernetes is all about running pods, it may be surprising

> to learn that there has been a loophole in securely providing images to pods for

Reviewer 3: ```suggestion
to learn that there has been a caveat to restricting pod access to authenticated images for
```

> over 10 years in the form of this [issue 18787](https://github.com/kubernetes/kubernetes/issues/18787)! 

Reviewer 3: ```suggestion
over 10 years in the form of [this issue](https://github.com/kubernetes/kubernetes/issues/18787)! 
```

> over 10 years in the form of [this issue](https://github.com/kubernetes/kubernetes/issues/18787)!

Reviewer 2: ```suggestion
over 10 years in the form of [issue 18787](https://github.com/kubernetes/kubernetes/issues/18787)!
```

It is an exciting release when you can resolve a ten year old issue.

{{< note >}}
Throughout this blog post, the term "pod credentials" will be used often. In this context,

> the term generally encapsulates the authentication materia that is available to a pod

Reviewer 3: ```suggestion
the term generally encapsulates the authentication material that is available to a pod
```

to authenticate a container image pull.
{{< /note >}}

## IfNotPresent, even if I'm not supposed to have it

> The gist of the problem is that `ImagePullPolicy: IfNotPresent` strategy has done

Reviewer 3: ```suggestion
The gist of the problem is that the `ImagePullPolicy: IfNotPresent` strategy has done
```

exactly its job. *Pod A* in a *Namespace X* is scheduled to a *Node 1*, and requires
*image Foo* from a private repository. The pod references *Secret 1* in its `imagePullSecrets`.
*Secret 1* contains the necessary credentials to pull from the private repository.
The Kubelet will utilize the credentials from *Secret 1* as supplied by *Pod A*
and it pulls *container image Foo* from the registry. This is exactly what you expect.

> But, if a *Pod B* in *Namespace Y* happens to also be scheduled to *Node 1*, funny things

Reviewer 3: ```suggestion
But, if *Pod B* in *Namespace Y* happens to also be scheduled to *Node 1*, funny things
```

> can happen. *Pod B* may require the same private image, specifying the `IfNotPreset`

Reviewer 3: ```suggestion
can happen. *Pod B* may reference the same private image, specifying the `IfNotPresent`
```

image pull policy. When the Kubelet tries to run the pod, it honors the `IfNotPresent`

> pulling policy, it sees that the *image Foo* is already present locally, and will

Reviewer 3: ```suggestion
policy, it sees that the *image Foo* is already present locally, and will
```

provide *image Foo* to *Pod B*. Surely that is ok since the container image is already
present, right?

{{< figure
    src="ensure_secret_image_pulls.svg"
    caption="Using a private image pulled by a different pod"
    alt="Illustration of the process of two pods trying to access a private image, the first one with a pull secret, the second one without it"
>}}

Wrong. While `IfNotPresent` should not pull *image Foo* if it is already present
on the node, it is an incorrect security posture to allow all pods scheduled
to a node to have access to that private image, discarding the fact that these
pods would otherwise never be able to pull that image from the remote repository
in the first place.

## IfNotPresent, but only if I am supposed to have it

> In Kubernetes 1.33, we are finally addressing this (really old) problem and

Reviewer 4: ```suggestion
In Kubernetes 1.33, we (SIG Auth and SIG Node) are finally addressing this (really old) problem and
```
Need to clarify who we is here

> getting the auth right! As expected, if an image is not present, the Kubelet will

Reviewer 3: ```suggestion
getting the verification right! As expected, if an image is not present, the Kubelet will
```

attempt to pull the image. The credentials each pod supplies will be utilized for this
task. Same as it has always been.

If the image is present, this is where the behavior changes.
The Kubelet will now verify the pod's credentials before allowing the pod to use
the image.

Performance and service stability have been a consideration while designing the feature.
Pods utilizing the same credential will not be required to re-authenticate. This is
also true in case the credentials sourced from the same Kubernetes Secret object rotated.

## Never pull, but prove you can use it!

> The `ImagePullPolicy: Never` option does not fetch images. Ever. However, if the

Reviewer 4: ```suggestion
The `imagePullPolicy: Never` option does not fetch images. Ever. However, if the
```

container image is already present on the node, the pod's credentials to use the
image may still require verification. 

Just like before, pods utilizing the same credential will not be required to re-authenticate.
If the pod is not able to supply credentials previously used to successfully pull an
image, though, the `Never` image pull policy will make it fail because without
the ability to pull images, the pod just has no way to access the image.

## Always?

> The `ImagePullPolicy: Always` has always worked as intended. Each time an image

Reviewer 4: ```suggestion
The `imagePullPolicy: Always` has always worked as intended. Each time an image
```

     they originated from.
  1. The Kubelet removes the original record about the intent to pull an image.

When a second pod comes, requesting the same private image that was pulled before:
  1. The Kubelet checks the credentials that the pod provided for the pull.
  1. If the credential hash, or the source Secret of the credentials match the hash or source
     Secret which were recorded for a successful pull before, the pod is allowed to use
     the local image.
  1. If the credentials or their source Secret are not found in the records of
     previous successful pulls for that image, the Kubelet will attempt to use
     these new credentials to request a pull from the remote registry.

## Try it out

In Kubernetes 1.33 we ship the alpha version of this feature. To give it a spin,
enable the `KubeletEnsureSecretPulledImages` feature gate for your 1.33 Kubelets.

You can learn more about the feature and how it is configured on the
[concept page for Images](/docs/concepts/containers/images/#ensureimagepullcredentialverification)
in the official Kubernetes documentation.

## What's next?

In future releases, there are a couple of areas for us to focus on:

> 1. Make the feature work together with [Projected service account tokens for Kubelet image credential providers](https://github.com/kubernetes/enhancements/issues/4412) which

Reviewer 4: nit

```suggestion
1. Make the feature work together with [Projected service account tokens for Kubelet image credential providers](https://kep.k8s.io/4412) which
```

>    add a new, workload-specific source of image pull credentials.

Reviewer 4: ```suggestion
   adds a new, workload-specific source of image pull credentials.
```

1. Write a benchmarking suite to be able to measure the performance of the feature,
   and to be able to measure the performance impact of any future changes to it.
1. Implement an in-memory caching layer so that we don't need to read files for each image pull request.

## How to get involved

> [Reading KEP-2535](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images)

Reviewer 4: nit

```suggestion
[Reading KEP-2535](https://kep.k8s.io/2535)
```

is a great way to understand these changes in depth.

> If you are interested in further involvement, reach out in Slack
> on the [#sig-auth-authenticators-dev](https://kubernetes.slack.com/archives/C04UMAUC4UA) channel. 

Reviewer 4: ```suggestion
If you are interested in further involvement, reach out to us
on the [#sig-auth-authenticators-dev](https://kubernetes.slack.com/archives/C04UMAUC4UA) channel
on Kubernetes Slack
(for an invitation, visit [https://slack.k8s.io/](https://slack.k8s.io/)).
```
